File: backend/src/api/routes/events.ts
import { Router, Request, Response } from 'express';
import { db } from '../../db/database';
import { logger } from '../../utils/logger';
import { eventQueueService } from '../../services/queue/EventQueueService';

const router = Router();

interface EventRow {
  id: string;
  type: string;
  equipment_id: string;
  equipment_name: string;
  payload: string;
  status: string;
  retry_count: number;
  next_retry_at: number | null;
  created_at: number;
  sent_at: number | null;
  error_message: string | null;
}

function mapEvent(row: EventRow) {
  return {
    id: row.id,
    type: row.type,
    equipmentId: row.equipment_id,
    equipmentName: row.equipment_name,
    payload: JSON.parse(row.payload) as Record<string, unknown>,
    status: row.status,
    retryCount: row.retry_count,
    nextRetryAt: row.next_retry_at || undefined,
    createdAt: row.created_at,
    sentAt: row.sent_at || undefined,
    errorMessage: row.error_message || undefined,
  };
}

// GET /api/events
router.get('/', (req: Request, res: Response) => {
  try {
    const { status, type, equipmentId } = req.query as Record<string, string | undefined>;
    const limit = Math.min(parseInt(String(req.query.limit || '100')), 500);
    const offset = parseInt(String(req.query.offset || '0'));

    const conditions: string[] = [];
    const params: (string | number)[] = [];

    if (status && status !== 'all') {
      conditions.push('status = ?');
      params.push(status.toUpperCase());
    }

    if (type && type !== 'all') {
      conditions.push('type = ?');
      params.push(type.toUpperCase());
    }

    if (equipmentId) {
      conditions.push('equipment_id = ?');
      params.push(equipmentId);
    }

    const where = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';

    const rows = db.prepare(`
      SELECT * FROM event_queue ${where} ORDER BY created_at DESC LIMIT ? OFFSET ?
    `).all(...params, limit, offset) as EventRow[];

    const countRow = db.prepare(`
      SELECT COUNT(*) as total FROM event_queue ${where}
    `).get(...params) as { total: number };

    res.json({
      events: rows.map(mapEvent),
      total: countRow.total,
    });
  } catch (err) {
    logger.error(`GET /events error: ${err instanceof Error ? err.message : String(err)}`);
    res.status(500).json({ error: 'Failed to fetch events' });
  }
});

// GET /api/events/stats
router.get('/stats', (_req: Request, res: Response) => {
  try {
    const stats = eventQueueService.getStats();

    const typeRows = db.prepare(`
      SELECT type, COUNT(*) as count FROM event_queue GROUP BY type
    `).all() as Array<{ type: string; count: number }>;

    const byType: Record<string, number> = { MACHINE_STATUS: 0, COUNTER: 0 };
    for (const row of typeRows) {
      byType[row.type] = row.count;
    }

    res.json({ ...stats, byType });
  } catch (err) {
    logger.error(`GET /events/stats error: ${err instanceof Error ? err.message : String(err)}`);
    res.status(500).json({ error: 'Failed to fetch event stats' });
  }
});

// DELETE /api/events/failed
router.delete('/failed', (_req: Request, res: Response) => {
  try {
    const result = db.prepare("DELETE FROM event_queue WHERE status = 'FAILED'").run();
    logger.info(`Cleared ${result.changes} failed events`);
    res.json({ deleted: result.changes });
  } catch (err) {
    logger.error(`DELETE /events/failed error: ${err instanceof Error ? err.message : String(err)}`);
    res.status(500).json({ error: 'Failed to clear failed events' });
  }
});

export default router;

File: backend/api/routes/events.py
import json

from fastapi import APIRouter, Query

from db import db_lock, get_db
from services.queue.event_queue_service import event_queue_service
from utils.logger import logger

router = APIRouter()


def _map_row(row: dict) -> dict:
    return {
        "id": row["id"],
        "type": row["type"],
        "equipmentId": row["equipment_id"],
        "equipmentName": row["equipment_name"],
        "payload": json.loads(row["payload"]),
        "status": row["status"],
        "retryCount": row["retry_count"],
        "nextRetryAt": row["next_retry_at"],
        "createdAt": row["created_at"],
        "sentAt": row["sent_at"],
        "errorMessage": row["error_message"],
    }


@router.get("")
def list_events(
    status: str | None = Query(default=None),
    type: str | None = Query(default=None),
    equipmentId: str | None = Query(default=None),
    limit: int = Query(default=100, le=500),
    offset: int = Query(default=0, ge=0),
):
    db = get_db()
    conditions = []
    params: list = []

    if status and status != "all":
        conditions.append("status = ?")
        params.append(status.upper())
    if type and type != "all":
        conditions.append("type = ?")
        params.append(type.upper())
    if equipmentId:
        conditions.append("equipment_id = ?")
        params.append(equipmentId)

    where = f"WHERE {' AND '.join(conditions)}" if conditions else ""

    rows = db.execute(
        f"SELECT * FROM event_queue {where} ORDER BY created_at DESC LIMIT ? OFFSET ?",
        (*params, limit, offset),
    ).fetchall()
    total = db.execute(f"SELECT COUNT(*) as total FROM event_queue {where}", params).fetchone()["total"]

    return {"events": [_map_row(dict(r)) for r in rows], "total": total}


@router.get("/stats")
def get_stats():
    stats = event_queue_service.get_stats()
    db = get_db()
    type_rows = db.execute(
        "SELECT type, COUNT(*) as count FROM event_queue GROUP BY type"
    ).fetchall()
    by_type = {"MACHINE_STATUS": 0, "COUNTER": 0}
    for row in type_rows:
        by_type[row["type"]] = row["count"]
    return {**stats, "byType": by_type}


@router.delete("/failed")
def clear_failed_events():
    db = get_db()
    with db_lock():
        cursor = db.execute("DELETE FROM event_queue WHERE status = 'FAILED'")
        db.commit()
    logger.info(f"Cleared {cursor.rowcount} failed events")
    return {"deleted": cursor.rowcount}
